WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift

```diff
@@ -0,0 +1,189 @@
+///
+enum ProductDownloadFileError: Error {
+    case emptyFileName
+    case emptyFileUrl
+    case invalidFileUrl
+}
+
+/// Provides view data for downloadable file, and handles init/UI/navigation actions needed in product downloadable file settings.
+///
+final class ProductDownloadFileViewModel: ProductDownloadFileViewModelOutput {
+    private let product: ProductFormDataModel
+    private let downloadableFileIndex: Int?
+
+    // Editable data
+    //
+    private(set) var fileName: String?
+    private(set) var fileURL: String?
+    private(set) var fileID: String?
+    private(set) var formType: ProductDownloadFileViewController.FormType
+
+    init(product: ProductFormDataModel,
+         downloadFileIndex: Int?,
+         formType: ProductDownloadFileViewController.FormType) {
+        self.product = product
+        self.downloadableFileIndex = downloadFileIndex
+        self.formType = formType
+
+        if let downloadableFileIndex = downloadableFileIndex,
+            let file = product.downloadableFiles[safe: downloadableFileIndex] {
+            fileName = file.name
+            fileURL = file.fileURL
+            fileID = file.downloadID
+        }
+    }
+
+    var sections: [Section] {
+        let nameSection = Section(footer: Strings.fileNameFooter, rows: [.name])
+        let urlSection = Section(footer: Strings.urlFooter, rows: [.url])
+
+        switch product {
+        case is EditableProductModel:
+            return [
+                urlSection,
+                nameSection
+            ]
+        default:
+            fatalError("Unsupported product type: \(product)")
+        }
+    }
+}
+
+// MARK: - UI changes
+//
+extension ProductDownloadFileViewModel: ProductDownloadFileActionHandler {
+
+    func handleFileNameChange(_ fileName: String?, onValidation: @escaping (_ isValid: Bool) -> Void) {
+        self.fileName = fileName
+
+        onValidation(isChangesValid())
+    }
+
+    func handleFileUrlChange(_ fileURL: String?, onValidation: @escaping (_ isValid: Bool) -> Void) {
+        self.fileURL = fileURL
+
+        onValidation(isChangesValid())
+    }
+
+    // MARK: - Navigation actions
+
+    func completeUpdating(onCompletion: ProductDownloadFileViewController.Completion) {
+        if let fileURL = fileURL, isChangesValid() {
+            onCompletion(fileName, fileURL, fileID, hasUnsavedChanges())
+        }
+        return
+    }
+
+    func hasUnsavedChanges() -> Bool {
+        return isChangesValid()
+    }
+}
+
+// MARK: - Convenience Methods
+//
+private extension ProductDownloadFileViewModel {
+    func isChangesValid() -> Bool {
+        var fileUrlChanged = false
+        var fileNameChanged = false
+
+        guard var newFileURL = self.fileURL,
+            newFileURL.isEmpty == false,
+            newFileURL.isValidURL() == true else {
+                return false
+        }
+        newFileURL = newFileURL.addHTTPSSchemeIfNecessary()
```

Comment:
I'm sorry but it's not clear to me why you are adding the https scheme here (altering the original URL).

Reply:
As per the last suggestion here: https://tinyurl.com/y6em2rzf,
It says `URL has to be a valid URL, if the user doesn’t enter a protocol, we’ll add “http” implicitly.`
That's why I made that change. But honestly, I think we shouldn't do it. So removing it. Please advise if I should add back it.